src/java/org/apache/cassandra/cql3/selection/Selection.java

```diff
@@ -187,7 +187,9 @@ public static Selection fromSelectors(TableMetadata table,
                                                                             factories,
                                                                             isJson);
 
-        return (processesSelection(selectables) || selectables.size() != selectedColumns.size() || hasGroupBy)
+        boolean hasMaskedColumns = selectedColumns.stream().anyMatch(ColumnMetadata::isMasked);
```

Comment:
Shall we avoid streams for perf reasons and move to a classic loop?

Reply:
This is run only once per query, so not in the hot path. It's relatively common to use streams while preparing the query.

Reply:
I disagree it's not in the hot path. It's not for a _specific query_ as yes, it gets prepared only once. But it's in the hot path in terms of how many total prepare requests we can handle. So in low prepared query scenarios _reusage_ it will have an impact.

IOW prepare once and query 1K times we're ok. Prepare once and reuse a handful of times we're not ok. High env traffic env where the prepared stmnts cache rotates often we're not ok. Nodes handling multi-tenant traffic may be handling a high number of prepares given the many different loads/apps they may support and we're not ok.

Maybe food for another ticket? I would still avoid it here if possible.

Reply:
I understand that the hot path where we avoid streams consists on that areas of the code that are repeated for every row or, more probably, column. The number of times those are hit is several orders of magnitude greater than in things that are used only to prepare the query, thereof the preference for performance over readability.

It seems to me that at the moment we are not avoiding streams in query building. Just taking a quick look at the CQL package there are multiple usages of streams, for example:
* [o.a.c.cql3/CQL3Type.java#L840](https://github.com/apache/cassandra/blob/cassandra-4.1.0/src/java/org/apache/cassandra/cql3/CQL3Type.java#L840)
* [o.a.c.cql3/CQL3Type.java#L876](https://github.com/apache/cassandra/blob/cassandra-4.1.0/src/java/org/apache/cassandra/cql3/CQL3Type.java#L876)
* [o.a.c.cql3/Lists.java#L133](https://github.com/apache/cassandra/blob/cassandra-4.1.0/src/java/org/apache/cassandra/cql3/Lists.java#L133)
* [o.a.c.cql3/Sets.java#L123](https://github.com/apache/cassandra/blob/cassandra-4.1.0/src/java/org/apache/cassandra/cql3/Sets.java#L123)
* [o.a.c.cql3/MultiColumnRelation.java#L225](https://github.com/apache/cassandra/blob/cassandra-4.1.0/src/java/org/apache/cassandra/cql3/MultiColumnRelation.java#L225)
* [o.a.c.cql3/QueryProcessor.java#L481](https://github.com/apache/cassandra/blob/cassandra-4.1.0/src/java/org/apache/cassandra/cql3/QueryProcessor.java#L481)
* [o.a.c.cql3/TokenRelation.java#L138](https://github.com/apache/cassandra/blob/cassandra-4.1.0/src/java/org/apache/cassandra/cql3/TokenRelation.java#L138)
* [o.a.c.cql3/conditions/ColumnConditions.java#L75](https://github.com/apache/cassandra/blob/cassandra-4.1.0/src/java/org/apache/cassandra/cql3/conditions/ColumnConditions.java#L75)
* [o.a.c.cql3/functions/AbstractFunction.java#L69](https://github.com/apache/cassandra/blob/cassandra-4.1.0/src/java/org/apache/cassandra/cql3/functions/AbstractFunction.java#L69)
* [o.a.c.cql3/functions/FunctionResolver.java#L203](https://github.com/apache/cassandra/blob/cassandra-4.1.0/src/java/org/apache/cassandra/cql3/functions/FunctionResolver.java#L203)
* [o.a.c.cql3/functions/UDAggregate.java#L97](https://github.com/apache/cassandra/blob/cassandra-4.1.0/src/java/org/apache/cassandra/cql3/functions/UDAggregate.java#L97)
* [o.a.c.cql3/restrictions/StatementRestrictions.java#L380](https://github.com/apache/cassandra/blob/cassandra-4.1.0/src/java/org/apache/cassandra/cql3/restrictions/StatementRestrictions.java#L380)
* [o.a.c.cql3/selection/AbstractFunctionSelector.java#L72](https://github.com/apache/cassandra/blob/cassandra-4.1.0/src/java/org/apache/cassandra/cql3/selection/AbstractFunctionSelector.java#L72)
* [o.a.c.cql3/selection/MapSelector.java#L113](https://github.com/apache/cassandra/blob/cassandra-4.1.0/src/java/org/apache/cassandra/cql3/selection/MapSelector.java#L113)
* [o.a.c.cql3/selection/Selectable.java#L708](https://github.com/apache/cassandra/blob/cassandra-4.1.0/src/java/org/apache/cassandra/cql3/selection/Selectable.java#L708)
* [o.a.c.cql3/selection/Selectable.java#L793](https://github.com/apache/cassandra/blob/cassandra-4.1.0/src/java/org/apache/cassandra/cql3/selection/Selectable.java#L793)
* [o.a.c.cql3/selection/Selectable.java#L886](https://github.com/apache/cassandra/blob/cassandra-4.1.0/src/java/org/apache/cassandra/cql3/selection/Selectable.java#L886)
* [o.a.c.cql3/selection/Selectable.java#L954](https://github.com/apache/cassandra/blob/cassandra-4.1.0/src/java/org/apache/cassandra/cql3/selection/Selectable.java#L954)
* [o.a.c.cql3/selection/Selectable.java#L1024](https://github.com/apache/cassandra/blob/cassandra-4.1.0/src/java/org/apache/cassandra/cql3/selection/Selectable.java#L1024)
* [o.a.c.cql3/statements/SelectStatement.java#L1278](https://github.com/apache/cassandra/blob/cassandra-4.1.0/src/java/org/apache/cassandra/cql3/statements/SelectStatement.java#L1278)

We should also check if queries are calling packages other than `o.a.c.cql3` that also use streams.

I'd say that getting rid of those streams used on query building leans towards premature optimization. But, if we are concerned about the performance of streams on query building we should probably open a followup ticket to audit where streams are used.

Reply:
I don't think readability is impacted if streams are dropped, they would become quite simple loops. Given it's a cheap and immediate optimization I would still avoid streams here if it were my PR but I'll defer to your preference being the author. It's something that in multi tenancy env might be a thing and will make a nice LHF ticket. I'll be happy to create it at least for discussion [CASSANDRA-18248](https://issues.apache.org/jira/browse/CASSANDRA-18248)

Reply:
As I try to argue in CASSANDRA-18248 I think it boils down to deciding if we see streams avoidance as a free lunch or not. Profiling would mean getting the actual change in iiuc.